# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2023-08-16T14:36:01.926121563+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuyce9fvn/model/ with prefix e2ec13ea101e4700
[INFO 2023-08-16T14:36:01.938295171+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:36:01.938656604+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2023-08-16T14:36:06.58007662+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4rju4svo/model/ with prefix 6b77cd64049b4267
[INFO 2023-08-16T14:36:06.589591817+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2023-08-16T14:36:06.589922915+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmpv3dzkfym as temporary training directory


[INFO 2023-08-16T14:38:14.057174208+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv3dzkfym/model/ with prefix a00096801a554237
[INFO 2023-08-16T14:38:14.071926019+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2023-08-16T14:38:14.071974854+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2023-08-16T14:38:14.07200182+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2023-08-16T14:38:31.304456902+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr6f7_hc5/model/ with prefix 1c3559c9050e4ae1
[INFO 2023-08-16T14:38:31.30925631+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:38:31.309336234+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2023-08-16T14:38:32.847107458+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx_sg9173/model/ with prefix c4d2a2a3cecc4f7f
[INFO 2023-08-16T14:38:32.867327728+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2023-08-16T14:38:33.923809212+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4sdgbidk/model/ with prefix 56261469212d4379
[INFO 2023-08-16T14:38:33.928614391+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2023-08-16T14:38:35.510781107+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4nan9iu6/model/ with prefix 0239889c15574522
[INFO 2023-08-16T14:38:35.538433735+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2023-08-16T14:38:36.601723434+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9j_ix9kk/model/ with prefix 8891b452bd1a4d1f
[INFO 2023-08-16T14:38:36.607724721+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2023-08-16T14:38:37.598384245+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjy7calck/model/ with prefix 70c8c77fef184fcb
[INFO 2023-08-16T14:38:37.601493213+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2023-08-16T14:38:38.708622621+00:00 kernel.cc:1214] Loading model from path /tmp/tmppmz3zk98/model/ with prefix 833a189270624c2d
[INFO 2023-08-16T14:38:38.716636692+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2023-08-16T14:38:40.189058173+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdwd213uq/model/ with prefix cfd5f5942efd4dfe
[INFO 2023-08-16T14:38:40.210162853+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2023-08-16T14:38:41.40162987+00:00 kernel.cc:1214] Loading model from path /tmp/tmp92wxzlua/model/ with prefix c1b8a04589994d25
[INFO 2023-08-16T14:38:41.412249726+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:38:41.412322389+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2023-08-16T14:38:43.202240379+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkxpqg699/model/ with prefix 231bbd6390b84197
[INFO 2023-08-16T14:38:43.219551412+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2023-08-16T14:38:44.306287287+00:00 kernel.cc:1214] Loading model from path /tmp/tmplubus_p8/model/ with prefix e742f0ca43774211
[INFO 2023-08-16T14:38:44.312706822+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2023-08-16T14:38:45.610177837+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7zzftlii/model/ with prefix 8c092a4955e74c8e
[INFO 2023-08-16T14:38:45.625915338+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2023-08-16T14:38:47.091513885+00:00 kernel.cc:1214] Loading model from path /tmp/tmpc3go5_rr/model/ with prefix 43565000ceeb4415
[INFO 2023-08-16T14:38:47.098053366+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2023-08-16T14:38:48.301832421+00:00 kernel.cc:1214] Loading model from path /tmp/tmphcxfv2zy/model/ with prefix 8c3b878bc34d44c2
[INFO 2023-08-16T14:38:48.313610213+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2023-08-16T14:38:49.401691835+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmoo0tsfy/model/ with prefix 9fc34526575f4a63
[INFO 2023-08-16T14:38:49.408232515+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2023-08-16T14:38:50.522511645+00:00 kernel.cc:1214] Loading model from path /tmp/tmpje_41h7m/model/ with prefix a397c9f9d6f14cb9
[INFO 2023-08-16T14:38:50.530419585+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2023-08-16T14:38:51.797213679+00:00 kernel.cc:1214] Loading model from path /tmp/tmpebxn_k03/model/ with prefix 9c207cc692224dac
[INFO 2023-08-16T14:38:51.81077588+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:38:51.810843047+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2023-08-16T14:38:53.555049788+00:00 kernel.cc:1214] Loading model from path /tmp/tmpi58zz_w9/model/ with prefix 98a4b5d2cf494e23
[INFO 2023-08-16T14:38:53.569595687+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2023-08-16T14:38:54.811775291+00:00 kernel.cc:1214] Loading model from path /tmp/tmp94zuc_z4/model/ with prefix dc981aa6190b4302
[INFO 2023-08-16T14:38:54.823833462+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2023-08-16T14:38:56.269102904+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk9mudlan/model/ with prefix ac7f63c55d784a07
[INFO 2023-08-16T14:38:56.289370524+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2023-08-16T14:38:57.625646821+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv9341hci/model/ with prefix ceeb26fd959b4e56
[INFO 2023-08-16T14:38:57.641872008+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2023-08-16T14:38:59.070751763+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbvb8rg9n/model/ with prefix c4f1cc3f829d448c
[INFO 2023-08-16T14:38:59.076471365+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2023-08-16T14:39:00.130505387+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmzx1tydf/model/ with prefix 359ab3f38ee44707
[INFO 2023-08-16T14:39:00.136611113+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2023-08-16T14:39:01.284130346+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb8r3uvhq/model/ with prefix 3a269f1f5dbc4205
[INFO 2023-08-16T14:39:01.29380068+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2023-08-16T14:39:02.96692726+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6apwopnv/model/ with prefix 24c2e70ed8954d38
[INFO 2023-08-16T14:39:02.973566982+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:39:02.973606079+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2023-08-16T14:39:04.845419948+00:00 kernel.cc:1214] Loading model from path /tmp/tmpux_xo8m0/model/ with prefix 359d9d7c37464857
[INFO 2023-08-16T14:39:04.857748277+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2023-08-16T14:39:06.494952866+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjil8yt21/model/ with prefix ababf0672de8492e
[INFO 2023-08-16T14:39:06.505553122+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2023-08-16T14:39:07.577414635+00:00 kernel.cc:1214] Loading model from path /tmp/tmpiq6axp29/model/ with prefix 8ec8da1d742446fa
[INFO 2023-08-16T14:39:07.584514347+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2023-08-16T14:39:08.64724329+00:00 kernel.cc:1214] Loading model from path /tmp/tmpemwhftbg/model/ with prefix 7841aa30dc7e4411
[INFO 2023-08-16T14:39:08.653075084+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2023-08-16T14:39:09.921747978+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzgdiz77o/model/ with prefix 03b825501bbb40db
[INFO 2023-08-16T14:39:09.93545838+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2023-08-16T14:39:11.99779871+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr3qey2zd/model/ with prefix cc7dee65b08e47ca
[INFO 2023-08-16T14:39:12.026275291+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2023-08-16T14:39:13.811047698+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_zjl0k0x/model/ with prefix 8d76c42a2ea94150
[INFO 2023-08-16T14:39:13.822940982+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:39:13.822986077+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2023-08-16T14:39:14.937003582+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbtg4xcdj/model/ with prefix bbc88f75d14b42aa
[INFO 2023-08-16T14:39:14.943584452+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2023-08-16T14:39:16.229380368+00:00 kernel.cc:1214] Loading model from path /tmp/tmpti964dtm/model/ with prefix f078a415dee24536
[INFO 2023-08-16T14:39:16.243821436+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2023-08-16T14:39:17.848435229+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsvt7xq6g/model/ with prefix 01b44fa93fd64f82
[INFO 2023-08-16T14:39:17.857874598+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2023-08-16T14:39:18.976403308+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq0vnlos7/model/ with prefix d98758b1252c4eab
[INFO 2023-08-16T14:39:18.985804079+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2023-08-16T14:39:20.277924817+00:00 kernel.cc:1214] Loading model from path /tmp/tmpds_nu1q2/model/ with prefix a86cfd3e2c904f32
[INFO 2023-08-16T14:39:20.293245329+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2023-08-16T14:39:21.448203114+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyuwzzb4v/model/ with prefix 40ebc48e154049be
[INFO 2023-08-16T14:39:21.458238145+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2023-08-16T14:39:22.76545516+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3obze_cc/model/ with prefix a22c7cea57a441f0
[INFO 2023-08-16T14:39:22.783026696+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2023-08-16T14:39:24.047111507+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuihgvwgp/model/ with prefix f15d2463516347ea
[INFO 2023-08-16T14:39:24.060848507+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:39:24.061117136+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2023-08-16T14:39:25.095560796+00:00 kernel.cc:1214] Loading model from path /tmp/tmpn0xv2t75/model/ with prefix 355a811cb5824088
[INFO 2023-08-16T14:39:25.100405602+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2023-08-16T14:39:26.436253777+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo40i05z1/model/ with prefix 21640fed765d488e
[INFO 2023-08-16T14:39:26.452640496+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2023-08-16T14:39:27.639682571+00:00 kernel.cc:1214] Loading model from path /tmp/tmptzspfvm7/model/ with prefix f42e1b7fcbb04488
[INFO 2023-08-16T14:39:27.649730473+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2023-08-16T14:39:29.035486564+00:00 kernel.cc:1214] Loading model from path /tmp/tmpz00mwcgu/model/ with prefix 7371faf2e09243c3
[INFO 2023-08-16T14:39:29.053663586+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2023-08-16T14:39:30.245277402+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7jg3et2o/model/ with prefix 690e5a9db8e349b7
[INFO 2023-08-16T14:39:30.255258566+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2023-08-16T14:39:31.291235495+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqts_6rwi/model/ with prefix 95ef4877f0294ea8
[INFO 2023-08-16T14:39:31.295033096+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2023-08-16T14:39:32.63151818+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4dxcrv30/model/ with prefix 07563d565bcf4b2e
[INFO 2023-08-16T14:39:32.646951255+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2023-08-16T14:39:33.953579143+00:00 kernel.cc:1214] Loading model from path /tmp/tmpusna3wk9/model/ with prefix 161ee5722aaa4d36
[INFO 2023-08-16T14:39:33.968057978+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2023-08-16T14:39:35.049658042+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqg99p1u_/model/ with prefix 89d9685adf524e35
[INFO 2023-08-16T14:39:35.05622445+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:39:35.056273686+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2023-08-16T14:39:36.198883261+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq93zlkb7/model/ with prefix f583c759d0f448cb
[INFO 2023-08-16T14:39:36.206085877+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2023-08-16T14:39:37.466135954+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwhcdfexw/model/ with prefix d80b4c1595aa4710
[INFO 2023-08-16T14:39:37.479157064+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2023-08-16T14:39:38.836896678+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuixwbvp_/model/ with prefix 97a155afee6541d1
[INFO 2023-08-16T14:39:38.853977214+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2023-08-16T14:39:40.006092094+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcoafvhgz/model/ with prefix 071b898172fc49b1
[INFO 2023-08-16T14:39:40.014158008+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2023-08-16T14:39:41.146791101+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe3uebzph/model/ with prefix d2f17d9d145f43df
[INFO 2023-08-16T14:39:41.154629549+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2023-08-16T14:39:42.179123532+00:00 kernel.cc:1214] Loading model from path /tmp/tmp04u478ge/model/ with prefix a969691b958c45b9
[INFO 2023-08-16T14:39:42.182764305+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2023-08-16T14:39:43.521974334+00:00 kernel.cc:1214] Loading model from path /tmp/tmpv096bgzh/model/ with prefix e29a304d549945e7
[INFO 2023-08-16T14:39:43.537438241+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2023-08-16T14:39:45.377407443+00:00 kernel.cc:1214] Loading model from path /tmp/tmprjsyns7m/model/ with prefix 475f0136b9694197
[INFO 2023-08-16T14:39:45.391167932+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:39:45.391216092+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2023-08-16T14:39:46.48748816+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoxz8uotl/model/ with prefix db2a163c28f84f26
[INFO 2023-08-16T14:39:46.492251841+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2023-08-16T14:39:47.631436589+00:00 kernel.cc:1214] Loading model from path /tmp/tmpul7td5wk/model/ with prefix f915dc8797eb4a83
[INFO 2023-08-16T14:39:47.638909934+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2023-08-16T14:39:48.854682569+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl3xnfslr/model/ with prefix 3eff66f7af964abc
[INFO 2023-08-16T14:39:48.864764206+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2023-08-16T14:39:50.074270161+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqkh0__58/model/ with prefix 34de3570079b4ec2
[INFO 2023-08-16T14:39:50.084972717+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2023-08-16T14:39:51.157965753+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw0hs5d6c/model/ with prefix 6c08fc230bd749fb
[INFO 2023-08-16T14:39:51.163505269+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2023-08-16T14:39:52.857429+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcrpiwz7c/model/ with prefix 52bf89fa968340ed
[INFO 2023-08-16T14:39:52.884530617+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2023-08-16T14:39:54.127676178+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9yp93aqv/model/ with prefix 121e2c5ded234111
[INFO 2023-08-16T14:39:54.137772356+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2023-08-16T14:39:55.276371346+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe0fwhm59/model/ with prefix 22e23fb09b1e4cff
[INFO 2023-08-16T14:39:55.284597797+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2023-08-16T14:39:56.34096924+00:00 kernel.cc:1214] Loading model from path /tmp/tmpap4do8mi/model/ with prefix 2ef7b8b254224bd2
[INFO 2023-08-16T14:39:56.34660818+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:39:56.346665644+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2023-08-16T14:39:57.450742813+00:00 kernel.cc:1214] Loading model from path /tmp/tmphemuu5vt/model/ with prefix fbdaf67dc0fd4315
[INFO 2023-08-16T14:39:57.45751128+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2023-08-16T14:39:58.887184938+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqtz8bzkw/model/ with prefix 266de6af396948cc
[INFO 2023-08-16T14:39:58.907319306+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2023-08-16T14:40:00.461243627+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy5uan1f7/model/ with prefix 9b587fb1e013444d
[INFO 2023-08-16T14:40:00.474015492+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2023-08-16T14:40:02.075789785+00:00 kernel.cc:1214] Loading model from path /tmp/tmpidfc6xq6/model/ with prefix 6526ceceec65430a
[INFO 2023-08-16T14:40:02.083024569+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2023-08-16T14:40:03.289407949+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuad9jx3e/model/ with prefix bd167f2994ce468c
[INFO 2023-08-16T14:40:03.299158499+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2023-08-16T14:40:04.452439099+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_cuh28ud/model/ with prefix a9265f77952547fc
[INFO 2023-08-16T14:40:04.460633768+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2023-08-16T14:40:06.535163162+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzuw4g3j4/model/ with prefix 86219456af744b76
[INFO 2023-08-16T14:40:06.551949553+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:40:06.551998985+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2023-08-16T14:40:07.695168743+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbro0z8ie/model/ with prefix 812b0d97a8024255
[INFO 2023-08-16T14:40:07.703394821+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2023-08-16T14:40:08.960348509+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoxcyn_i_/model/ with prefix 486c60c07e6d4663
[INFO 2023-08-16T14:40:08.973000062+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2023-08-16T14:40:10.614821587+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1g9xn6q9/model/ with prefix d2adb7679529484e
[INFO 2023-08-16T14:40:10.623452921+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2023-08-16T14:40:12.269485976+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg4lrkaq9/model/ with prefix c2be82afd92c4b2e
[INFO 2023-08-16T14:40:12.273962539+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2023-08-16T14:40:13.389959286+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3p2gcz3m/model/ with prefix 09675642951748e4
[INFO 2023-08-16T14:40:13.395250607+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2023-08-16T14:40:14.502108134+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9n6cvv0i/model/ with prefix f9a80647f3454717
[INFO 2023-08-16T14:40:14.509244244+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2023-08-16T14:40:15.641000144+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg6ub9m37/model/ with prefix 47e87487ceff4d6d
[INFO 2023-08-16T14:40:15.648824576+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2023-08-16T14:40:17.468389006+00:00 kernel.cc:1214] Loading model from path /tmp/tmpubnfnzlm/model/ with prefix 3309f70fc9b24e5f
[INFO 2023-08-16T14:40:17.478952489+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:40:17.478997098+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2023-08-16T14:40:18.723626167+00:00 kernel.cc:1214] Loading model from path /tmp/tmppd_4morj/model/ with prefix 119bef82f0ec40a1
[INFO 2023-08-16T14:40:18.735421927+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2023-08-16T14:40:19.928335391+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgodzoyww/model/ with prefix e746d8a60aaa4e75
[INFO 2023-08-16T14:40:19.93878996+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2023-08-16T14:40:21.133353154+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjelg5ezt/model/ with prefix ffa7110a51e8430a
[INFO 2023-08-16T14:40:21.142477104+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2023-08-16T14:40:23.324042934+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjst7mtos/model/ with prefix c6a3a264662e4ae0
[INFO 2023-08-16T14:40:23.343875776+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2023-08-16T14:40:24.53409332+00:00 kernel.cc:1214] Loading model from path /tmp/tmprqx3clsi/model/ with prefix bb3dadd4e3544adc
[INFO 2023-08-16T14:40:24.541587027+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2023-08-16T14:40:26.567826871+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgm70oncv/model/ with prefix 8221bb9ec95b4e5e
[INFO 2023-08-16T14:40:26.585918138+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2023-08-16T14:40:28.427696365+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwq2wn53z/model/ with prefix 401632924b19471e
[INFO 2023-08-16T14:40:28.447429272+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:40:28.447492881+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2023-08-16T14:40:29.742190535+00:00 kernel.cc:1214] Loading model from path /tmp/tmpipwpmcv6/model/ with prefix 2d655e5f7178400b
[INFO 2023-08-16T14:40:29.755187853+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2023-08-16T14:40:30.855721847+00:00 kernel.cc:1214] Loading model from path /tmp/tmpur_pf7ds/model/ with prefix 89acfb641bb24cf4
[INFO 2023-08-16T14:40:30.860964598+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2023-08-16T14:40:32.119038287+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkratgsd3/model/ with prefix cfe85a75276d4d6b
[INFO 2023-08-16T14:40:32.130365+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2023-08-16T14:40:33.308379194+00:00 kernel.cc:1214] Loading model from path /tmp/tmpml4pjc_4/model/ with prefix 0e780d15a57d44f3
[INFO 2023-08-16T14:40:33.315573371+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2023-08-16T14:40:35.355089138+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6e8pg1mc/model/ with prefix 15d7b6c1ee934a04
[INFO 2023-08-16T14:40:35.37382418+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2023-08-16T14:40:36.590356594+00:00 kernel.cc:1214] Loading model from path /tmp/tmpy0qzpjan/model/ with prefix eeadae20d4e34fed
[INFO 2023-08-16T14:40:36.600997522+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2023-08-16T14:40:38.278058931+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6hoxdybn/model/ with prefix b8dcad2e1caf49e3
[INFO 2023-08-16T14:40:38.285015789+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2023-08-16T14:40:39.481473583+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw_b4gte5/model/ with prefix c4860a9c74a54502
[INFO 2023-08-16T14:40:39.49042827+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-08-16T14:40:39.490472436+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2023-08-16T14:40:40.722723988+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_8gh8ngb/model/ with prefix dc084c4daf9a437b
[INFO 2023-08-16T14:40:40.733341176+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2023-08-16T14:40:42.365190877+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl1aewncg/model/ with prefix 4f5aa125c145476c
[INFO 2023-08-16T14:40:42.370936161+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2023-08-16T14:40:43.602026014+00:00 kernel.cc:1214] Loading model from path /tmp/tmplzohdwqq/model/ with prefix f220fc14ec804d79
[INFO 2023-08-16T14:40:43.610097068+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2023-08-16T14:40:44.955836254+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd65vtn6e/model/ with prefix 632527426d794d5a
[INFO 2023-08-16T14:40:44.971865489+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 